components/omega/doc/design/Decomp.md

```diff
@@ -0,0 +1,260 @@
+local operators. The minimum is typically 3 for current
+algorithms. 
+
+These parameters and the mesh input file will be read as part of
+the input configuration file in a decomp configuration group:
+
+    decomp:
+       meshInputFilename: 'omegaMeshFile.nc'
+       partitionMethod: 'MetisKWay'
+       haloWidth: 3
+
+The Metis KWay partition method will be default and the haloWidth
+will initially default to 3.
+
+#### 4.1.2 Class/structs/data types
+
+There will be a Decomp class that stores and defines a partitioning
+of the address space. 
+
+    class Decomp {
+
+       private:
+          partMethod method;   ///< method to use for partitioning
+
+
+       public:
+
+          I4 haloDepth;    ///< depth of the halo
+
+          I4 nCellsGlobal; ///< num cells in the global domain
+          I4 nCellsAll;    ///< num cells on local rank (own+halos)
+          I4 nCellsOwned;  ///< num cells exclusively owned by local rank
+
+          std::vector<I4>[] nCellsHalo[]; ///< num of owned+halo points
+                                          ///< for each level of halo
+
+          std::vector<I4>[] cellGlobalID[]; ///< global index for each
+                                            ///< local cell (own+halo)
+
+          /// The adjacency (neighbor) info will be stored in the
+          /// compact form that Metis uses. In this form the neighbor
+          /// cell location for each owned and ghost cell in local cell
+          /// order is stored as a linear vector of integer pairs
+          /// (partition/processor number and local address). A second
+          /// vector stores the starting address in this list for the
+          /// neighbors associated with the local cell address.
+
+          std::vector<I4>[] nbrStartCell ///< start addr in nbr array for cell
+          std::vector<I4>[] nbrLocCell   ///< list of nbr addresses
+
+          [ repeat identical variables for edge and vertex index spaces ]
+ 
+
+          // methods described below
+
+    }
+
+### 4.2 Methods
+
+Because most class members are public, only constructors and
+destructors are really needed.
+
+#### 4.2.1 Constructor
+
+The main method will be a constructor that decomposes the mesh and
+creates the public variables. The default decomposition will require
+no arguments and will read options from the input config file:
+
+    Decomp mainDecomp;
+
+For multiple decomposition of the same mesh, a second constructor
+will be needed that starts from the default decomposition and
+uses arguments for determining decomposition options. The exact form
+will be determined later, but might look something like:
+
+    Decomp newDecomp(oldDecomp, nParts, method, haloWidth);
+
+#### 4.2.2 Destructor
+
+After most of the initialization is complete and the decomposition
+information has been used to initialize meshes, halo, I/O, etc.,
+we will supply a destructor to release the space in memory.
+
+    delete myDecomp;
+
+
+## 5 Verification and Testing
```

Comment:
We could add a unit test that the sum of local cells equals the global number, and similarly for edges and vertices. 

We could also do a more complicated sum check across processors to ensure partitions are correct. For example, we could sum the global index of owned cells (edges, vertices), and also sum their associated x,y,z or lat,lon coordinates. That sum should be identical for any number of partitions, including one (i.e. not partitioned). The sum would be exact for integers, and within order of operations errors for floating point values.